### PR TITLE
feat(vulnerabilities): Improve vulnerabilities overview page to speed up loading time

### PR DESCRIPTION
**Pull request by bs-jokri**
https://github.com/sw360/sw360portal/issues/411
https://github.com/sw360/sw360portal/issues/574

### CODE REVIEW

**Comment by alexbrdn** [comment]
@bs-jokri, @bodetc, did you notice that travis build failed? please have a look into it,

**Comment by mcjaeger** [comment]
@bs-jokri  two, eh more actually, questions:
1) how about changing commits to conventional changelog (and remove the issue numbers)
2) the history does not look OK, rebase instead of merge, ...
4) travis ...
5) make fully qualified issue references in the PR comment?

**Comment by maxhbr** [comment]
I think this PR contains version problems since the travis log in line https://travis-ci.org/sw360/sw360portal/jobs/272469001#L2898 mentions version 2.1.0 of SW360:
```
 [WARNING] The POM for org.eclipse.sw360:src-common:jar:2.1.0-SNAPSHOT is missing, no dependency information available
```

**Comment by maxhbr** [comment]
The problem is https://github.com/sw360/sw360portal/pull/564/files#diff-81a3c9e205dbaebba467187e2c21fe67R42, please use the version variable instead of a fixed version

**Comment by mcjaeger** [comment]
(from discussion) it would be good to have an example vulnerability file.

**Comment by bs-jokri** [comment]
@bodetc Filter key should stay in the input field if filter is applied (same behaviour as in components service)

**Comment by maxhbr** [comment]
On my system the following error appears in the browser log after accessing the component detail page:

But I do not think that this is related to your changes. I think that we are missing a requireJS again.

File: frontend/sw360-portlet/src/main/webapp/html/vulnerabilities/view.jsp [hunk 1]
```diff
@@ -34,9 +45,61 @@
...
+                <th class="infoheading">
+                    Loading
+                </th>
+            </tr>
+            </thead>
+            <tbody style="background-color: #f8f7f7; border: none;">
+            <tr>
+                <td>
+                    <select class="searchbar" id="view_size" name="<portlet:namespace/><%=PortalConstants.VIEW_SIZE%>" onchange="reloadViewSize()">
+                        <option value="50" <core_rt:if test="${viewSize == 50}">selected</core_rt:if>>50 latest</option>
```
Comment (maxhbr) [inline frontend/sw360-portlet/src/main/webapp/html/vulnerabilities/view.jsp]: I think we can remove that option (later).

**Comment by maxhbr** [comment]
Works as expected.

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/PortletUtils.java [hunk 2]
```diff
@@ -158,7 +137,7 @@ public static ECCStatus getEccStatusFromString(String enumNumber) {
         }
     }
 
-    public static <U extends TFieldIdEnum> Object enumFromString(String value, U field) {
+    private static <U extends TFieldIdEnum> Object enumFromString(String value, U field) {
```
Comment (maxhbr) [inline frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/PortletUtils.java]: Minor comment: unnecessary change.

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/vulnerabilities/VulnerabilitiesPortlet.java [hunk 3]
```diff
@@ -92,11 +144,11 @@ private void shortenTimeStampsToDates(List<Vulnerability> vulnerabilities){
         });
     }
 
-    private String getDateFromFormattedTimeStamp(String formattedTimeStamp){
-        return formattedTimeStamp.substring(0,10);
+    private String getDateFromFormattedTimeStamp(String formattedTimeStamp) {
+        return formattedTimeStamp.substring(0, 10);
```
Comment (maxhbr) [inline frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/vulnerabilities/VulnerabilitiesPortlet.java]: Magic constant

File: libraries/lib-datahandler/src/main/thrift/vulnerabilities.thrift [hunk 4]
```diff
@@ -218,4 +234,16 @@ service VulnerabilityService {
       * returns null otherwise
       **/
     Vulnerability getVulnerabilityByExternalId(1: string externalId, 2: User user);
+
+    /**
+      * returns the vulnerability with given externalId if it exists in the database and if user is valid
+      * returns null otherwise
+      **/
+    list<Vulnerability> getVulnerabilitiesByExternalIdOrConfiguration(1: optional string externalId, 2: optional string vulnerableConfiguration, 3: User user);
```
Comment (maxhbr) [inline libraries/lib-datahandler/src/main/thrift/vulnerabilities.thrift]: `optional` arguments in a function call? I do not think that we use that somewhere else.

File: frontend/sw360-portlet/src/main/webapp/html/vulnerabilities/view.jsp (hunk 1 above)
Comment (bodetc) [reply frontend/sw360-portlet/src/main/webapp/html/vulnerabilities/view.jsp]: done

File: backend/src/src-vulnerabilities/src/main/java/org/eclipse/sw360/vulnerabilities/VulnerabilityHandler.java [hunk 5]
```diff
@@ -213,62 +213,95 @@ private VulnerabilityDTO enrichVulnerabilityDTO(VulnerabilityDTO dto, String rel
...
+        if (!PermissionUtils.isUserAtLeast(UserGroup.USER, user)) {
             return null;
         }
         return dbHandler.getByExternalId(Vulnerability.class, externalId);
     }
 
     @Override
+    public VulnerabilityWithReleaseRelations getVulnerabilityWithReleaseRelationsByExternalId(String externalId, User user) throws TException {
+        if (!PermissionUtils.isUserAtLeast(UserGroup.USER, user)) {
+            return null;
```
[the remark this answers is not in the document]
Comment (bodetc) [reply backend/src/src-vulnerabilities/src/main/java/org/eclipse/sw360/vulnerabilities/VulnerabilityHandler.java]: We should then change that consistently everywhere...

File: backend/src/src-vulnerabilities/src/main/java/org/eclipse/sw360/vulnerabilities/db/VulnerabilityRepository.java [hunk 6]
```diff
@@ -35,16 +36,18 @@
...
                     "  } " +
                     "}";
 
-    private static final String ALL_EXTERNALIDS =
-            "function(doc) {" +
-                    "  if (doc.type == 'vulnerability') {" +
-                    "    emit(null, doc.externalId);" +
-                    "  } " +
+    private static final String BY_VULNERABLE_CONFIGURATION =
+            "function(doc) {\n" +
```
[the remark this answers is not in the document]
Comment (bodetc) [reply backend/src/src-vulnerabilities/src/main/java/org/eclipse/sw360/vulnerabilities/db/VulnerabilityRepository.java]: Removed

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/CustomFieldHelper.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
+package org.eclipse.sw360.portal.common;
```
[the remark this answers is not in the document]
Comment (bodetc) [reply frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/CustomFieldHelper.java]: Done.

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/vulnerabilities/VulnerabilitiesPortlet.java (hunk 3 above)
Comment (bodetc) [reply frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/vulnerabilities/VulnerabilitiesPortlet.java]: In a magic function, I don't think it brings anything to extract it in this case :)

File: libraries/lib-datahandler/src/main/thrift/vulnerabilities.thrift (hunk 4 above)
Comment (bodetc) [reply libraries/lib-datahandler/src/main/thrift/vulnerabilities.thrift]: We do not, indeed, but since they are not required together, why not?

**Comment by maxhbr** [comment]
On the vulnerability detail page in the footer one sees for example the following:

where `Release Name` is the name of the release, but the link behind it links to its component **which might have a different name**. That is not a common situation and one might want to ignore that comment.

**Comment by maxhbr** [comment]
@mcjaeger This PR works as expected.

**Comment by bs-jokri** [comment]
@bodetc this should have been fixed already

**Comment by bodetc** [comment]
Yes, but I think it got lost in one rebase or merge... Since I am working on another vulnerability story right now, I can fix it there. Unless there is more open comments on this PR.
